pulsar4s-akka-streams/src/main/scala/com/sksamuel/pulsar4s/akka/streams/PulsarCommittableSourceGraphStage.scala

```diff
@@ -57,36 +86,13 @@ class PulsarCommittableSourceGraphStage[T](create: () => Consumer[T], seek: Opti
     override def onPull(): Unit = {
       implicit val context: ExecutionContext = super.materializer.executionContext
       logger.debug("Pull received; asking consumer for message")
-
-      consumer.receiveAsync.onComplete {
-        case Success(msg) =>
-          logger.debug(s"Message received: $msg")
-          receiveCallback.invoke(new CommittableMessage[T] {
-            override def message: ConsumerMessage[T] = msg
-            override def ack(cumulative: Boolean): Future[Done] = {
-              logger.debug(s"Acknowledging message: $msg")
-              val ackFuture = if (cumulative) {
-                consumer.acknowledgeCumulativeAsync(msg.messageId)
-              } else {
-                consumer.acknowledgeAsync(msg.messageId)
-              }
-              ackFuture.map(_ => Done)
-            }
-            override def nack(): Future[Done] = {
-              logger.debug(s"Negatively acknowledging message: $msg")
-              consumer.negativeAcknowledgeAsync(msg.messageId).map(_ => Done)
-            }
-          })
-        case Failure(e) =>
-          logger.warn("Error when receiving message", e)
-          failStage(e)
-      }
+      consumer.receiveAsync.onComplete(receiveCallback.invoke)
     }
 
-    override def stop(): Unit = completeStage()
+    override def stop(): Unit = stopCallback.invoke(())
 
     override def shutdown()(implicit ec: ExecutionContext): Future[Done] = {
-      completeStage()
+      stop()
```

Comment:
Here we invoke stop, which invokes stopCallback, which is async based, and we don't do anything with the returned async promise. So we just immediately go into the next line. I think it was probably correct before?

Reply:
I invoked `stop()` here because it invokes the callback, which we want to do here since `shutdown` is being called from a different thread. Alternatively, we could invoke `stopCallback` directly here, but I don't think it makes sense to call `completeStage` directly.

Reply:
I guess it might not be clear is that `stop` is not part of the lifecycle of the Akka `GraphStageLogic`, but a method on the `Control`. This class implements both so they can share state, but it would probably be clearer to split them up into separate objects. I could do that now or in a separate pull request.

Reply:
Okay, I see. So you're right that `stop` is async, so we need to wait for the callback to finish before we shut down the consumer. But the code as-is isn't correct since `shutdown` is being invoked from a different thread and `completeStage` accesses internal state.

Reply:
The simplest solution I could think of is to complete a promise when the stage is stopped then flatMap on the promise. Unfortunately we can't call `stopCallback.invokeWithFeedback`, because the feedback promise can't be completed after the stage is shut down.

Another minor difference now is that `stop()` no longer guarantees that the operator is stopped when it returns, since it's calling an async callback. But the only way to fix that is to await, which I don't think makes sense. It would make more sense to have an API where `stop()` returns a `Future[Done]`.

Reply:
Yes agreed, but I think its acceptable for stop to block, given that it's an inherited method.
I would add another - closeAsync or stopAsync or whatever that can return a Future[Done].

Reply:
In that case I'm wondering if we should just deprecate stop and add a new async method. I don't like the idea of having a blocking API in an Akka application without making it obvious that it's a blocking API.